cmd/fake-lepton/main.go

```diff
@@ -126,23 +129,41 @@ func connectToSocket(dbusService *service) error {
 }
 
 func sendCPTV(conn *net.UnixConn, file string) error {
-    mu.Lock()
-    defer mu.Unlock()
-    fullpath := path.Join(cptvDir, file)
-    if _, err := os.Stat(fullpath); err != nil {
-        fmt.Printf("%v does not exist\n", fullpath)
-        return err
-    }
-    log.Printf("sending raw frames of %v\n", fullpath)
-    r, err := cptv.NewFileReader(fullpath)
-    if err != nil {
-        return err
+    select {
+    case lockChannel <- struct{}{}:
+        fullpath := path.Join(cptvDir, file)
+        if _, err := os.Stat(fullpath); err != nil {
+            log.Printf("%v does not exist\n", fullpath)
+            unlock()
+            return err
+        }
+        r, err := cptv.NewFileReader(fullpath)
+        if err != nil {
+            unlock()
+            return err
+        }
+
+        log.Printf("sending raw frames of %v\n", fullpath)
+        go SendFrames(conn, r)
+        return nil
+    case <-time.After(lockTimeout):
+        // lock not acquired
+        return errors.New("Could not acquire lock")
     }
+
+}
+
+func unlock() {
+    <-lockChannel
+}
+func SendFrames(conn *net.UnixConn, r *cptv.FileReader) error {
     defer r.Close()
```

Comment:
Similarly the defer r.Close() should be called just after you open r

Reply:
Similarly because of the go routine this would close it before it is read from. The reason it is opened before entering send frames is so that some usefull feed back can be sent back to the webserver. If there is an error opening the file.